File: 125. Valid Palindrome/Main.java
class Main {
    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.isPalindrome(" a#bA"));
        System.out.println(solution.isPalindrome("0P"));
        System.out.println(solution.isPalindrome(" "));
        System.out.println(solution.isPalindrome(""));
    }
}

class Solution {
    public boolean isPalindrome(String s) {
        s = s.toLowerCase();
        int left = 0, right = s.length() - 1;

        // discard non-alphanumeric characters
        while (left < s.length() && !isAlphnumeric(s.charAt(left))) left++;
        while (right >= 0 && !isAlphnumeric(s.charAt(right))) right--;

        while (left <= right) {
            // check equality
            if (s.charAt(left) != s.charAt(right)) {
                return false;
            }

            // move pointers
            left++;
            right--;

            // discard non-alphanumeric characters
            while (left < s.length() && !isAlphnumeric(s.charAt(left))) left++;
            while (right >= 0 && !isAlphnumeric(s.charAt(right))) right--;
        }

        return true;
    }

    private boolean isAlphnumeric(char c) {
        if ((c >= '0' && c <= '9') || (c >= 'a' && c <= 'z')) {
            return true;
        } else {
            return false;
        }
    }
}

File: 125. Valid Palindrome/main.go
package main

import "fmt"

func main() {
	fmt.Println(isPalindrome(" a#bA"))
	fmt.Println(isPalindrome("0P"))
	fmt.Println(isPalindrome(" "))
	fmt.Println(isPalindrome(""))
}

func isPalindrome(s string) bool {
	left := 0
	right := len(s) - 1

	for left <= right {
		if !isAlphanumeric(rune(s[left])) {
			left++
		} else if !isAlphanumeric(rune(s[right])) {
			right--
		} else {
			if uppercase(rune(s[left])) != uppercase(rune(s[right])) {
				return false
			} else {
				left++
				right--
			}
		}
	}

	return true
}

func uppercase(r rune) rune {
	if r >= 'a' && r <= 'z' {
		return r - 32
	} else {
		return r
	}
}

func isAlphanumeric(r rune) bool {
	if (r >= '0' && r <= '9') || (r >= 'a' && r <= 'z') || (r >= 'A' && r <= 'Z') {
		return true
	} else {
		return false
	}
}
